### PR TITLE
Updated all pagination queries to work with the `Stream` instead of `Iterator`

### PR DESCRIPTION
**Pull request by xgreenx**
This change updates all pagination queries to work with the async stream instead of the sync iterator.

It is preparation for the next PR to move the fetching of transactions, coins, and messages into a separate thread and await the return of the result from this thread. 

Also, this PR limits the maximum number of requested coins from the `coins_to_spend` query to `max_inputs`.

## Checklist
- [x] Breaking changes are clearly marked as such in the PR description and changelog

### Before requesting review
- [x] I have reviewed the code myself

### CODE REVIEW

File: crates/fuel-core/src/query/balance.rs [hunk 1]
```diff
@@ -28,29 +29,28 @@ use std::{
...
+    pub async fn balance(
         &self,
         owner: Address,
         asset_id: AssetId,
         base_asset_id: AssetId,
     ) -> StorageResult<AddressBalance> {
         let amount = AssetQuery::new(
             &owner,
-            &AssetSpendTarget::new(asset_id, u64::MAX, usize::MAX),
+            &AssetSpendTarget::new(asset_id, u64::MAX, u16::MAX),
```
Comment (rymnc) [inline crates/fuel-core/src/query/balance.rs:40]: there's quite a bit of difference between usize::MAX (18446744073709551615) and u16::MAX (65535)
nice optimization

File: crates/fuel-core/src/query/balance.rs [hunk 2]
```diff
@@ -28,29 +29,28 @@ use std::{
...
-
-            // Increase the balance
-            balance = balance.saturating_add(amount);
-
-            Ok(balance)
-        })?;
+        .try_fold(0u64, |balance, amount| {
+            async move {
+                // Increase the balance
+                Ok(balance.saturating_add(amount))
```
Comment (rymnc) [inline crates/fuel-core/src/query/balance.rs:50]: this behaviour would be incorrect if the user has a balance of > 18.45 trillion (possible for memecoin projects/?)

Comment (rymnc) [reply crates/fuel-core/src/query/balance.rs:50]: but it's a major edge case so feel free to ignore

File: crates/fuel-core/src/query/balance/asset_query.rs [hunk 3]
```diff
@@ -196,7 +213,14 @@ impl<'a> AssetQuery<'a> {
 
     /// Returns the iterator over all valid(spendable, allowed by `exclude`) coins of the `owner`
     /// for the `asset_id`.
-    pub fn coins(&self) -> impl Iterator<Item = StorageResult<CoinType>> + '_ {
+    pub fn coins(self) -> impl Stream<Item = StorageResult<CoinType>> + 'a {
         self.query.coins()
     }
 }
+
+fn has_asset(assets: &Option<HashSet<&AssetId>>, asset_id: &AssetId) -> bool {
```
Comment (rymnc) [inline crates/fuel-core/src/query/balance/asset_query.rs:221]: hmm, maybe the signature of this function should be -
```suggestion
fn has_asset(assets: &HashSet<&AssetId>, asset_id: &AssetId) -> bool {
```

and leave it to the caller to decide whether to call or not based on `assets` having a value inside the `Option`.

File: crates/fuel-core/src/query/balance/asset_query.rs [hunk 4]
```diff
@@ -196,7 +213,14 @@ impl<'a> AssetQuery<'a> {
...
+    assets
+        .as_ref()
+        .map(|assets| assets.contains(asset_id))
+        .unwrap_or(true)
```
Comment (rymnc) [inline crates/fuel-core/src/query/balance/asset_query.rs:225]: shouldn't this be false if the Option is `None`?

File: crates/storage/src/kv_store.rs [hunk 5]
```diff
@@ -17,13 +17,8 @@ use core::ops::Deref;
...
 /// The key of the storage.
 pub type Key = Vec<u8>;
-#[cfg(feature = "std")]
 /// The value of the storage. It is wrapped into the `Arc` to provide less cloning of massive objects.
-pub type Value = std::sync::Arc<Vec<u8>>;
-
-#[cfg(not(feature = "std"))]
-/// The value of the storage. It is wrapped into the `Rc` to provide less cloning of massive objects.
-pub type Value = alloc::rc::Rc<Vec<u8>>;
+pub type Value = alloc::sync::Arc<Vec<u8>>;
```
Comment (rymnc) [inline crates/storage/src/kv_store.rs:21]: is this change intended to be in this PR? if so, should we use `alloc::sync::Arc` from this point onwards instead of stdlib?

**Review by rymnc — COMMENTED** [review]
left a few comments and suggestions :)

File: crates/storage/src/kv_store.rs (hunk 5 above)
Comment (xgreenx) [reply crates/storage/src/kv_store.rs:21]: `alloc::sync::Arc` and `std::sync::Arc` are the same in the context of `std`. Yes, without this change, it is not possible to compile code because I require `Send` on all `BoxedStreams` now

File: crates/fuel-core/src/graphql_api/database.rs [hunk 6]
```diff
@@ -306,28 +310,35 @@ impl ReadView {
...
+
+        futures::stream::iter(iter)
     }
 
-    pub fn owned_message_ids(
-        &self,
-        owner: &Address,
+    pub fn owned_message_ids<'a>(
+        &'a self,
+        owner: &'a Address,
```
[the remark this answers is not in the document]
Comment (xgreenx) [reply crates/fuel-core/src/graphql_api/database.rs:321]: Have you tried to compile with the suggested change? I remember before it was failing. Plus usually Rust says, when we can remove lifetime during ocmpilation

File: crates/fuel-core/src/query/balance.rs (hunk 2 above)
Comment (xgreenx) [reply crates/fuel-core/src/query/balance.rs:50]: It is how it worked before=)

File: crates/fuel-core/src/graphql_api/database.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (rymnc) [reply crates/fuel-core/src/graphql_api/database.rs:321]: yup tried it and ran cargo check too

[the remark this answers is not in the document]
Comment (rymnc) [reply crates/fuel-core/src/graphql_api/database.rs:321]: image

File: crates/fuel-core/src/graphql_api/database.rs [hunk 7]
```diff
@@ -306,28 +310,35 @@ impl ReadView {
...
         start_message_id: Option<Nonce>,
         direction: IterDirection,
-    ) -> BoxedIter<'_, StorageResult<Nonce>> {
-        self.off_chain
-            .owned_message_ids(owner, start_message_id, direction)
+    ) -> impl Stream<Item = StorageResult<Nonce>> + 'a {
```
[the remark this answers is not in the document]
Comment (xgreenx) [reply crates/fuel-core/src/graphql_api/database.rs:324]: Done in follow up PR https://github.com/FuelLabs/fuel-core/pull/2340/commits/d982fbc2ba14f3ac72249c0fad7ec129db33f727

File: crates/fuel-core/src/graphql_api/database.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (xgreenx) [reply crates/fuel-core/src/graphql_api/database.rs:321]: Done in follow up PR https://github.com/FuelLabs/fuel-core/pull/2340/commits/d982fbc2ba14f3ac72249c0fad7ec129db33f727

File: crates/fuel-core/src/query/balance/asset_query.rs (hunk 3 above)
Comment (xgreenx) [reply crates/fuel-core/src/query/balance/asset_query.rs:221]: https://github.com/FuelLabs/fuel-core/pull/2341/commits/c355cd1569c785379018912aa3ab1a1661d7ee8b

File: crates/fuel-core/src/query/balance/asset_query.rs (hunk 4 above)
Comment (xgreenx) [reply crates/fuel-core/src/query/balance/asset_query.rs:225]: https://github.com/FuelLabs/fuel-core/pull/2341/commits/c355cd1569c785379018912aa3ab1a1661d7ee8b
